core/service/pool.go

```diff
@@ -173,5 +178,25 @@ func (i *Instance) Proposal() market.ServiceProposal {
 
 // State returns the service instance state.
 func (i *Instance) State() State {
+	i.stateLock.RLock()
+	defer i.stateLock.RUnlock()
 	return i.state
 }
+
+func (i *Instance) setState(newState State) {
```

Comment:
It's a bit weird that state switch is not related to event publishing. Because this is exact place where events happen.  I would expect some kind of SetState or SwitchState which will publish even immediately after state switch.

Reply:
moved publish to setState